tensorflow-framework/src/main/java/org/tensorflow/framework/metrics/impl/Reduce.java

```diff
@@ -0,0 +1,239 @@
+  protected Variable<T> total;
+  /** the variable that holds the count of the metric values */
+  protected Variable<T> count;
+
+  /**
+   * Creates a Reducible Metric with a metric reductions of {@link MetricReduction#SUM}
+   *
+   * @param tf the TensorFlow Ops
+   * @param name the name for this metric. If null, name defaults to {@link Class#getSimpleName()}.
+   * @param seed the seed for random number generation. An initializer created with a given seed
+   *     will always produce the same random tensor for a given shape and data type.
+   * @param type the type for the variables and result
+   */
+  protected Reduce(Ops tf, String name, long seed, Class<T> type) {
+    this(tf, name, MetricReduction.SUM, seed, type);
+  }
+
+  /**
+   * @param tf The TensorFlow Ops
+   * @param name the name for this metric. If null, name defaults to {@link Class#getSimpleName()}.
+   * @param reduction The type of metric reduction to apply
+   * @param seed the seed for random number generation. An initializer created with a given seed
+   *     will always produce the same random tensor for a given shape and data type.
+   * @param type the type for the variables and result
+   */
+  protected Reduce(Ops tf, String name, MetricReduction reduction, long seed, Class<T> type) {
+    super(tf, name, seed);
+    this.reduction = reduction;
+    this.totalName = this.getVariableName(TOTAL);
+    this.countName = this.getVariableName(COUNT);
+    this.type = type;
+    setupVars();
+  }
+  /** Initializes the Variables */
+  private void setupVars() {
+    if (total == null) {
+      total = getTF().withName(totalName).variable(Shape.scalar(), type);
+    }
+    if (reduction == MetricReduction.SUM_OVER_BATCH_SIZE
+        || reduction == MetricReduction.WEIGHTED_MEAN) {
+      if (count == null) {
+        count = getTF().withName(countName).variable(Shape.scalar(), type);
+      }
+    }
+  }
+
+  /** {@inheritDoc} */
+  public Op resetStates() {
+    List<Op> controls = new ArrayList<>();
+    if (total != null) {
+      controls.add(
+          getTF().assign(total, CastHelper.cast(getTF(), getTF().constant(0), total.type())));
+    }
+    if (count != null) {
+      controls.add(
+          getTF().assign(count, CastHelper.cast(getTF(), getTF().constant(0), count.type())));
+    }
+    return getTF().withControlDependencies(controls).noOp();
+  }
+
+  /**
+   * Updates the metric variables based on the inputs. At least one input arg required for <code>
+   * values</code>, an optional additional input for the <code>sampleWeights</code>
+   *
+   * @param values the inputs to be passed to update state, this may not be null
+   * @param sampleWeights sample weights to be applied to values, may be null.
+   * @return the result with a control dependency on update state Operands
+   * @throws IllegalArgumentException if values is null
+   */
+  @Override
+  public List<Op> updateStateList(Operand<U> values, Operand<T> sampleWeights) {
```

Comment:
Does it usually help us to type `sampleWeights` as an `Operand<T>`, instead of letting it have its own tensor type? It feels arbitrary to me, and I see it doesn't help us in this case.

Reply:
I don't quite understand the question.   Did you mean SampleWeights be typed to something like Operand<Z> where <Z extends TNumber> so that sample weights is not tied to the Metric class type <T extends TNumber>?

Reply:
Right. Is it helpful that the tensor type of `sampleWeights` is tied to the `Metric` result type `T`, instead of either being method-local or being tied to the tensor type `U` of `values`?

Reply:
Eventually, all the types get cast to the Metric type `T` for the result and the variables, but there is no reason sampleWeights could not be another independent type. I will change the class signature.

Reply:
👍 done